Fixes for #1161, #1285 (Create), #1278, #1279

zaafonin opened the pull request:
<img width="854" height="480" alt="2025-07-31_15 34 57" src="https://github.com/user-attachments/assets/8dccff81-17d9-45ed-9a02-a7ccd5c21560" />

The reason for #1161 (encased fans on ships failing to bulk process recipes that are triggered by putting a full block in front of a fan, like a blaze burner or fan catalysts from addons): VS2 raycasts to check if a fan is obstructed by any ship. This raycast does not exclude fan's own ship, so if 1) a fan is on a ship and 2) there's a full block in front of a fan it, the air flow will stop. Now we first check obstructions in the Create way (same ship), then check for ship obstruction while excluding own ship.

Also bulk processing world items by ship fans was broken in 2.5, probably because of misported mixins. Now these mixins are version-agnostic so fans work again both on v0.5.1 and v6.

For the second commit: afaik @ThePlasticPotato said it's ok to uncomment these lines and it was already done in some other branch. Correct me if I'm wrong.

zaafonin committed:
Uncomment Create in Forge runtime, it works just fine

zaafonin committed:
Fix for #1161 + fixed regressions introduced when Create got dual-versioned

TechTastic reviewed (changes requested):
While this does fix Fans working on Ships, it seems the processing abilities do not affect items in worldspace from a Ship
<img width="1564" height="642" alt="image" src="https://github.com/user-attachments/assets/230e936f-fc5b-4a18-9128-561e800665ff" />

It also seems to not work on scaled Ships
<img width="1266" height="874" alt="image" src="https://github.com/user-attachments/assets/7f5bdc2b-8279-4305-a9ab-31af580f4337" />

It also seems that Ships block fans too early
<img width="1120" height="730" alt="image" src="https://github.com/user-attachments/assets/f9c9e57d-437d-4891-aba7-d380fe561cba" />

*Tbh, these last two issues are likely the same issue*

TechTastic commented:
<img width="906" height="694" alt="image" src="https://github.com/user-attachments/assets/7104b97a-c7cd-445b-bc07-0062e5725b23" />

Ok, so it *does* work on scaled Ships but without a block close to it and a LOT closer to the fans

zaafonin committed:
Processing now works on scaled ships

zaafonin commented:
Should be fixed now. Fan strength and distance is scaled matching the ship scale, but the distance is consistent so bulk processing works. There might be rare problems with processing worldspace items on miniships as the air current AABB is smaller and there's a chance for an item to miss it completely (at least that's how I understand it), but most of the time it's fine. Depots work as well

zaafonin committed:
I hope this time I really have solved encased fans...

zaafonin commented:
https://github.com/user-attachments/assets/bea39d6a-9fbd-44e4-87bc-8127d071ddd9

Now everything should work on tilted ships, including particles.

zaafonin committed:
Ok now encased fans should work as expected, even on tilted ships

zaafonin committed:
I can't believe stepOn and updateEntityAfterFallOn didn't work on ships because of an ambiguous mixin target reference

zaafonin committed:
Better detection for whether an ejector should transform its ejection vector from ship to world

zaafonin commented:
Fix for #1285, also version-independent.

https://github.com/user-attachments/assets/a90fae84-2096-4e0b-8536-e5cda44c1f4a

zaafonin committed:
Improved obstruction detection for building on ships, allows safe pillaring

zaafonin committed:
Fixed cobwebs, flowers and other blocks without a proper VoxelShape crashing the game when placed on ships (my own fault)

zaafonin committed:
Fixed funnels yet again

zaafonin committed:
Fixed remap = false included by mistake (cherrypickable into the good Create branch)

zaafonin committed:
Entities can interact with shipyard fluids on Fabric. Not my fix, it just got untracked in ...mixins.json for some reason

zaafonin commented:
Summary for this PR as asked:
- #1161: fixed, parts of MixinAirCurrent were rewritten so obstructions are handled correctly. Also fixed processing on scaled ships
- #1285: fixed, the culprit was Entity#stepOn not firing when entities stepped on ships. Coincidentally this fixes #1279 (sneaking on ships gives extra speed) and probably other bugs arising from players not technically stepping on ship blocks.
- #1278: fixed, now placement checks work for worldspace players and shipspace blocks. This uses an OBB intersection check (worldspace player AABB against shipspace block AABB with ship-to-world transformation) because a "regular" AABB transform would inflate it, yielding a lot of false positives. However, player's AABB is slightly deflated to allow some wiggle room: clientside and serverside ship transforms can be noticeably different, so a precise check can succeed on one side but fail on the other. An occasional false negative is bad but frequent ghost blocks are worse.
- Shipspace funnels barely ever accepting dropped items: fixed, now uses a simpler injection point, works on both Create versions.
- Entities not interacting with shipyard fluids on Fabric: fixed, the functional mixin was there but not enabled in config. Maybe it wasn't enabled for some obscure reason?

ThePlasticPotato commented:
lgtm
